task2_Yiska_Levi/src/Board.cpp

```diff
@@ -0,0 +1,293 @@
+        std::cout << "Invalid destination position." << std::endl;
+        return false;
+    }
+
+    // Check if the source position has a tool
+    GeneralTool* tool = _board[fromX][fromY];
+    if (tool == nullptr) {
+        return 11;
+    }
+    // Check if the tool in the source square is an opponent's tool
+    if (tool->isWhite()!= _white_turn) {
+        return 12;
+    }
+    //**
+    // Check if the tool can make the specified move
+    if (!tool->move(toX, toY, _board)) {
+        std::cout << "Invalid move for the selected tool." << std::endl;
+        return 21;
+    }
+
+    // Check if the destination position has a tool of the same color
+    if (_board[toX][toY] != nullptr && _board[toX][toY]->isWhite() == tool->isWhite()) {
+        return 13;
+    }
+
+    // Move the tool to the destination position
+    Board t(*this); // Create a copy of the current board
+    if (t._board[toX][toY] != NULL)
+        delete t._board[toX][toY];
+    t._board[toX][toY] = tool;
+    t._board[fromX][fromY] = nullptr;
+    //We will check that the action does not cause chess
+    if (isCheckmate(_white_turn)) {
+        return 31;
+    }
+
+    if (_board[toX][toY]!=NULL)
+        delete _board[toX][toY];
+    _board[toX][toY] = tool;
+    _board[fromX][fromY] = nullptr;
+    tool->setPosition(toX, toY);
+
+    //turn change 
+    if (_white_turn) {
+        _white_turn = false;
+    }
+    else {
+        _white_turn = true;
+    }
+    //Checking if the move caused a check on the opponent
+    if (isCheck(_white_turn))
+        return 41;
+    return 42;
+}
+
+bool Board::isKing(GeneralTool* a) {
+    return typeid(*a) == typeid(King);
+}
+
+std::pair<int, int> Board::findKing(bool isWhite) {
+    // Iterate over the board to find the king
+    for (int i = 0; i < 8; ++i) {
+        for (int j = 0; j < 8; ++j) {
+            GeneralTool* piece = _board[i][j];
+            if (piece != nullptr && isKing(piece) && piece->isWhite() == isWhite) {
+                // Found the king with the matching color
+                return std::make_pair(i, j);
+            }
+        }
+    }
+
+    // King not found, return invalid coordinates
+    return std::make_pair(-1, -1);
+
+}
+
+bool Board::isCheck(bool isWhiteKing)
+{
+    // Find the position of the king
+    std::pair<int, int> kingPosition = findKing(isWhiteKing);
+    int kingX = kingPosition.first;
+    int kingY = kingPosition.second;
+
+    // Check if the king position is valid
+    if (kingX == -1 || kingY == -1) {
+        // King not found, return false
+        return false;
+    }
+
+    // Iterate over opponent's pieces to check if they can attack the king
+    for (int x = 0; x < 8; x++) {
+        for (int y = 0; y < 8; y++) {
+            GeneralTool* tool = _board[x][y];
+            if (tool != nullptr && tool->isWhite() != isWhiteKing) {
+                std::vector<std::pair<int, int>> moves = tool->getPossibleMoves(_board);
```

Comment:
you are comparing the king with all possible move, and before that you calculate all the moves, isn't it better if you passed the position of the king and checked if the object can reach that position ?